In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('events_table.csv')
df.shape

(1952, 15)

### Preview the dataframe

In [3]:
df.head(5)

id                        created                                  uuid  \
0   96  2019-11-12 08:21:52.302787+00  20acd28b-4c0c-4760-8ef1-146b0d549415   
1   97  2019-11-12 08:21:53.329285+00  0242ac12-0005-fc3a-f03a-000000000007   
2   98  2019-11-12 08:21:53.325446+00  5a10185f-7412-4e8a-97c1-c8d71bd2aa4c   
3   99  2019-11-12 08:21:53.424271+00  0242ac12-0005-fc3a-f03a-00000000001f   
4  100  2019-11-12 08:21:53.425163+00  37b72c3c-a282-4a48-a77d-bbadb66a61c8   

                            parent_uuid                   event  \
0                                   NaN                 verbose   
1  5a10185f-7412-4e8a-97c1-c8d71bd2aa4c  playbook_on_play_start   
2                                   NaN       playbook_on_start   
3  0242ac12-0005-fc3a-f03a-000000000007  playbook_on_task_start   
4  0242ac12-0005-fc3a-f03a-00000000001f         runner_on_start   

      task_action failed changed        playbook                 play  \
0             NaN      f       f             NaN                  NaN   
1             NaN      f       f  using_role.yml  use galaxy git role   
2             NaN      f       t  using_role.yml                  NaN   
3  "gather_facts"      f       f  using_role.yml  use galaxy git role   
4  "gather_facts"      f       f  using_role.yml  use galaxy git role   

              task role  job_id  host_id host_name  
0              NaN  NaN       5      NaN       NaN  
1              NaN  NaN       5      NaN       NaN  
2              NaN  NaN       5      NaN       NaN  
3  Gathering Facts  NaN       5      NaN       NaN  
4  Gathering Facts  NaN       5      3.0  ansible1

### Add a new column to the dataframe -  `time_created`, which is a `datetime` object representation of the `created` column

In [4]:
df['time_created'] = df.apply(lambda row: pd.to_datetime(row['created']), axis=1)

### Verify that the column `time_created` has been added (scroll to the end)

In [5]:
df.head(5)

id                        created                                  uuid  \
0   96  2019-11-12 08:21:52.302787+00  20acd28b-4c0c-4760-8ef1-146b0d549415   
1   97  2019-11-12 08:21:53.329285+00  0242ac12-0005-fc3a-f03a-000000000007   
2   98  2019-11-12 08:21:53.325446+00  5a10185f-7412-4e8a-97c1-c8d71bd2aa4c   
3   99  2019-11-12 08:21:53.424271+00  0242ac12-0005-fc3a-f03a-00000000001f   
4  100  2019-11-12 08:21:53.425163+00  37b72c3c-a282-4a48-a77d-bbadb66a61c8   

                            parent_uuid                   event  \
0                                   NaN                 verbose   
1  5a10185f-7412-4e8a-97c1-c8d71bd2aa4c  playbook_on_play_start   
2                                   NaN       playbook_on_start   
3  0242ac12-0005-fc3a-f03a-000000000007  playbook_on_task_start   
4  0242ac12-0005-fc3a-f03a-00000000001f         runner_on_start   

      task_action failed changed        playbook                 play  \
0             NaN      f       f             NaN                  NaN   
1             NaN      f       f  using_role.yml  use galaxy git role   
2             NaN      f       t  using_role.yml                  NaN   
3  "gather_facts"      f       f  using_role.yml  use galaxy git role   
4  "gather_facts"      f       f  using_role.yml  use galaxy git role   

              task role  job_id  host_id host_name  \
0              NaN  NaN       5      NaN       NaN   
1              NaN  NaN       5      NaN       NaN   
2              NaN  NaN       5      NaN       NaN   
3  Gathering Facts  NaN       5      NaN       NaN   
4  Gathering Facts  NaN       5      3.0  ansible1   

                      time_created  
0 2019-11-12 08:21:52.302787+00:00  
1 2019-11-12 08:21:53.329285+00:00  
2 2019-11-12 08:21:53.325446+00:00  
3 2019-11-12 08:21:53.424271+00:00  
4 2019-11-12 08:21:53.425163+00:00

### We will be specifically doing our preliminary data analysis on `job_id=120`
### This is a Job that contains only 1 host in the inventory
### It installs Apache Server using ansible role

In [6]:
jobid = 120

### Known info: Job ID 120 was executed for a single host only, namely host `a1`

### Let's extract all the rows with `job_id=120`

In [7]:
job120 = df[df.job_id == jobid]
job120.shape

(62, 16)

### The above info reveals that job120 contains 62 rows or events

### Let's look at all the 62 events in about 3 batches (each with about 20-22 events) since Jupyter does not display the entire set by default

### First Group of Events - group1

In [8]:
group1 = job120[0:20]
group1

id                        created  \
1360  1456  2019-11-13 03:33:46.625738+00   
1361  1457  2019-11-13 03:33:48.085837+00   
1362  1458   2019-11-13 03:33:48.15501+00   
1363  1459  2019-11-13 03:33:48.163638+00   
1364  1460  2019-11-13 03:33:48.102049+00   
1365  1461  2019-11-13 03:33:50.632294+00   
1366  1462    2019-11-13 03:33:50.6332+00   
1367  1463  2019-11-13 03:33:50.586341+00   
1368  1464  2019-11-13 03:33:50.897341+00   
1369  1465  2019-11-13 03:33:50.871247+00   
1370  1466  2019-11-13 03:33:51.081304+00   
1371  1467  2019-11-13 03:33:50.898175+00   
1372  1468  2019-11-13 03:33:51.127379+00   
1373  1469  2019-11-13 03:33:51.128201+00   
1374  1470  2019-11-13 03:33:51.431555+00   
1375  1471  2019-11-13 03:33:51.488076+00   
1376  1472  2019-11-13 03:33:51.616439+00   
1377  1473  2019-11-13 03:33:51.489056+00   
1378  1474  2019-11-13 03:33:51.720417+00   
1379  1475  2019-11-13 03:33:51.675197+00   

                                      uuid  \
1360  c1641b85-a5c2-4949-a0f3-4362c130f031   
1361  00f6b9b4-67f7-41bf-ada8-e105bb3f0dd2   
1362  0242ac12-0005-3310-f85a-00000000001d   
1363  7909a668-c73f-48ae-be52-1282691a8213   
1364  0242ac12-0005-3310-f85a-000000000006   
1365  0242ac12-0005-3310-f85a-00000000000b   
1366  499598ee-4465-463c-9829-3f142bb2f7bc   
1367  88cec8fa-979e-4e5b-92d6-9ac89875bfde   
1368  0242ac12-0005-3310-f85a-00000000000c   
1369  c38b4686-261c-485d-a063-f2561102c9e5   
1370  a2079a20-f2b1-4d3d-acf0-1a98a9f48ea7   
1371  62e63ede-f1ea-44ce-824d-c28ef79627ec   
1372  0242ac12-0005-3310-f85a-00000000000d   
1373  55af2ff5-d01a-4cd5-9780-ca5f5702ae21   
1374  4feb0af1-2f98-452d-bbfb-e69758fca546   
1375  0242ac12-0005-3310-f85a-00000000000e   
1376  3ee137f7-52aa-416c-8db1-25f2bcc63e04   
1377  266bc4f0-23ea-4d6a-a7f0-4f027f510eaa   
1378  0242ac12-0005-3310-f85a-000000000047   
1379  99427293-2c63-4265-b970-1eb41da45f3a   

                               parent_uuid                   event  \
1360                                   NaN                 verbose   
1361                                   NaN       playbook_on_start   
1362  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1363  0242ac12-0005-3310-f85a-00000000001d         runner_on_start   
1364  00f6b9b4-67f7-41bf-ada8-e105bb3f0dd2  playbook_on_play_start   
1365  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1366  0242ac12-0005-3310-f85a-00000000000b         runner_on_start   
1367  0242ac12-0005-3310-f85a-00000000001d            runner_on_ok   
1368  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1369  0242ac12-0005-3310-f85a-00000000000b            runner_on_ok   
1370  0242ac12-0005-3310-f85a-00000000000c       runner_on_skipped   
1371  0242ac12-0005-3310-f85a-00000000000c         runner_on_start   
1372  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1373  0242ac12-0005-3310-f85a-00000000000d         runner_on_start   
1374  0242ac12-0005-3310-f85a-00000000000d            runner_on_ok   
1375  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1376  0242ac12-0005-3310-f85a-00000000000e            runner_on_ok   
1377  0242ac12-0005-3310-f85a-00000000000e         runner_on_start   
1378  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1379  0242ac12-0005-3310-f85a-00000000000e     playbook_on_include   

          task_action failed changed         playbook                    play  \
1360              NaN      f       f              NaN                     NaN   
1361              NaN      f       t  apache_role.yml                     NaN   
1362   "gather_facts"      f       f  apache_role.yml  use galaxy apache role   
1363   "gather_facts"      f       f  apache_role.yml  use galaxy apache role   
1364              NaN      f       f  apache_role.yml  use galaxy apache role   
1365   "include_vars"      f       f  apache_role.yml  use galaxy apache role   
1366   "include_vars"      f       f  apache_role.yml  us

### Second Group of Events - group2

In [9]:
group2 = job120[20:40]
group2

id                        created  \
1380  1476   2019-11-13 03:33:51.72133+00   
1381  1477  2019-11-13 03:34:03.440679+00   
1382  1478  2019-11-13 03:34:03.491893+00   
1383  1479  2019-11-13 03:34:03.492813+00   
1384  1480  2019-11-13 03:34:04.618952+00   
1385  1481   2019-11-13 03:34:04.61983+00   
1386  1482  2019-11-13 03:34:04.597248+00   
1387  1483  2019-11-13 03:34:04.905326+00   
1388  1484  2019-11-13 03:34:04.959603+00   
1389  1485  2019-11-13 03:34:05.287596+00   
1390  1486  2019-11-13 03:34:04.960378+00   
1391  1487  2019-11-13 03:34:05.288476+00   
1392  1488  2019-11-13 03:34:05.214172+00   
1393  1489  2019-11-13 03:34:05.574802+00   
1394  1490  2019-11-13 03:34:05.521229+00   
1395  1491  2019-11-13 03:34:05.584515+00   
1396  1492  2019-11-13 03:34:05.841024+00   
1397  1493  2019-11-13 03:34:06.061209+00   
1398  1494   2019-11-13 03:34:06.10663+00   
1399  1495  2019-11-13 03:34:06.107793+00   

                                      uuid  \
1380  f7fe8922-cd3b-4f6a-8cb5-0443487d8eee   
1381  06cc2c4f-de2d-4feb-b68e-f80cd30c7ea1   
1382  0242ac12-0005-3310-f85a-00000000000f   
1383  79c912f5-95a8-4511-988d-08e3414729f7   
1384  0242ac12-0005-3310-f85a-000000000010   
1385  8efbfa66-059f-4b93-b137-5d0e37e1c1d9   
1386  b48e2f7c-e993-409a-badd-aa6d27f8715a   
1387  5635ebcf-8c3c-4119-8f83-8332a3cfcb75   
1388  0242ac12-0005-3310-f85a-000000000011   
1389  0242ac12-0005-3310-f85a-000000000012   
1390  8dd0bedb-aecb-46ee-9886-08345d0d3495   
1391  4b66f1f0-abaa-4629-abb4-179a43741e86   
1392  abd88e15-7ca5-4d7c-8403-d5beba524805   
1393  0242ac12-0005-3310-f85a-000000000013   
1394  c896e5a2-0adf-4c78-b99c-136fbb5d0da6   
1395  db242b56-0761-4778-a664-ae467e10e9cf   
1396  75774e28-b79f-480c-a3dd-e1481a7d63e5   
1397  b40e042e-a090-43d7-8363-37fab94a3c01   
1398  0242ac12-0005-3310-f85a-000000000070   
1399  c07520c7-f298-4126-9cfe-c32cbd074111   

                               parent_uuid                   event  \
1380  0242ac12-0005-3310-f85a-000000000047         runner_on_start   
1381  0242ac12-0005-3310-f85a-000000000047            runner_on_ok   
1382  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1383  0242ac12-0005-3310-f85a-00000000000f         runner_on_start   
1384  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1385  0242ac12-0005-3310-f85a-000000000010         runner_on_start   
1386  0242ac12-0005-3310-f85a-00000000000f            runner_on_ok   
1387  0242ac12-0005-3310-f85a-000000000010            runner_on_ok   
1388  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1389  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1390  0242ac12-0005-3310-f85a-000000000011         runner_on_start   
1391  0242ac12-0005-3310-f85a-000000000012         runner_on_start   
1392  0242ac12-0005-3310-f85a-000000000011       runner_on_skipped   
1393  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1394  0242ac12-0005-3310-f85a-000000000012            runner_on_ok   
1395  0242ac12-0005-3310-f85a-000000000013         runner_on_start   
1396  0242ac12-0005-3310-f85a-000000000013            runner_on_ok   
1397  0242ac12-0005-3310-f85a-000000000013     playbook_on_include   
1398  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1399  0242ac12-0005-3310-f85a-000000000070         runner_on_start   

          task_action failed changed         playbook                    play  \
1380        "package"      f       f  apache_role.yml  use galaxy apache role   
1381        "package"      f       t  apache_role.yml  use galaxy apache role   
1382        "command"      f       f  apache_role.yml  use galaxy apache role   
1383        "command"      f       f  apache_role.yml  use galaxy apache role   
1384       "set_fact"      f       f  apache_role.yml  use galaxy apache role   
1385       "set_fact"      f       f  apache_role.yml  use galaxy apache role   
1386        "command"      f       f  apache_role.yml  us

### Third Group of Events - group3

In [10]:
group3 = job120[40:62]
group3

id                        created  \
1400  1496  2019-11-13 03:34:07.021707+00   
1401  1497   2019-11-13 03:34:07.02355+00   
1402  1498   2019-11-13 03:34:07.04765+00   
1403  1499  2019-11-13 03:34:07.235028+00   
1404  1500  2019-11-13 03:34:07.307896+00   
1405  1501  2019-11-13 03:34:07.308813+00   
1406  1502  2019-11-13 03:34:07.048882+00   
1407  1503   2019-11-13 03:34:09.11257+00   
1408  1504  2019-11-13 03:34:09.086828+00   
1409  1505  2019-11-13 03:34:09.090239+00   
1410  1506   2019-11-13 03:34:09.11328+00   
1411  1507  2019-11-13 03:34:09.313573+00   
1412  1508  2019-11-13 03:34:09.401317+00   
1413  1509  2019-11-13 03:34:09.402079+00   
1414  1510  2019-11-13 03:34:09.602825+00   
1415  1511  2019-11-13 03:34:09.725246+00   
1416  1512  2019-11-13 03:34:09.726231+00   
1417  1513  2019-11-13 03:34:11.557804+00   
1418  1514  2019-11-13 03:34:11.520995+00   
1419  1515  2019-11-13 03:34:11.523737+00   
1420  1516  2019-11-13 03:34:13.751631+00   
1421  1517  2019-11-13 03:34:13.749025+00   

                                      uuid  \
1400  800799f9-d0c7-47ce-a70f-f228647473b0   
1401  3577064c-dd28-4efb-97d1-4eb3cf800d15   
1402  0242ac12-0005-3310-f85a-000000000071   
1403  ac88ec15-8519-4fd0-8247-31b62c583296   
1404  0242ac12-0005-3310-f85a-000000000072   
1405  b0814b90-893e-46a0-80e1-f34ca24da692   
1406  0e49aff4-2b17-4936-a0c0-95331b47962c   
1407  0242ac12-0005-3310-f85a-000000000073   
1408  05092683-f47d-4ba3-a845-eaff1b146b11   
1409  912b57c5-c39b-4b20-9c2e-59c1a4681552   
1410  c1521c13-c9bc-4339-bbeb-e07eacc2955b   
1411  bed1b13c-9455-4394-987d-d4a0fd30d15d   
1412  0242ac12-0005-3310-f85a-000000000074   
1413  7203873d-aed0-4e46-9c5b-fbc8251bfac1   
1414  e4322f97-61a2-4dda-b99b-58935121e07e   
1415  0242ac12-0005-3310-f85a-000000000014   
1416  ef5eadf1-9755-41e5-b8c7-2c8519dce4ae   
1417  c8b9654a-c729-454a-b567-0bbf43bd28b8   
1418  71c064b8-9e0c-4f44-817f-498b794b6ee6   
1419  0242ac12-0005-3310-f85a-000000000016   
1420  dc2de65a-d372-426b-95a3-1c1d16f8504e   
1421  ce541cee-5bb6-4a55-bf6a-3891acf78ca0   

                               parent_uuid                   event  \
1400  0242ac12-0005-3310-f85a-000000000070       runner_item_on_ok   
1401  0242ac12-0005-3310-f85a-000000000070            runner_on_ok   
1402  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1403  0242ac12-0005-3310-f85a-000000000071       runner_on_skipped   
1404  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1405  0242ac12-0005-3310-f85a-000000000072         runner_on_start   
1406  0242ac12-0005-3310-f85a-000000000071         runner_on_start   
1407  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1408  0242ac12-0005-3310-f85a-000000000072      playbook_on_notify   
1409  0242ac12-0005-3310-f85a-000000000072            runner_on_ok   
1410  0242ac12-0005-3310-f85a-000000000073         runner_on_start   
1411  0242ac12-0005-3310-f85a-000000000073       runner_on_skipped   
1412  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1413  0242ac12-0005-3310-f85a-000000000074         runner_on_start   
1414  0242ac12-0005-3310-f85a-000000000074       runner_on_skipped   
1415  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1416  0242ac12-0005-3310-f85a-000000000014         runner_on_start   
1417  0242ac12-0005-3310-f85a-000000000016         runner_on_start   
1418  0242ac12-0005-3310-f85a-000000000014            runner_on_ok   
1419  0242ac12-0005-3310-f85a-000000000006  playbook_on_task_start   
1420  00f6b9b4-67f7-41bf-ada8-e105bb3f0dd2       playbook_on_stats   
1421  0242ac12-0005-3310-f85a-000000000016            runner_on_ok   

       task_action failed changed         playbook                    play  \
1400  "lineinfile"      f       f  apache_role.yml  use galaxy apache role   
1401  "lineinfile"      f       f  apache_role.yml  use galaxy apache role   
1402        "stat"      f       f  apache_role.yml  use galaxy apach

### Function that calculates the delta between 2 Time Strings

In [11]:
def time_elapsed(a, b):
    a = pd.to_datetime(a)
    b = pd.to_datetime(b)
    if b > a:
        c = b - a
    else:
        c = a - b
    c = c.total_seconds()
    return c

### Get all the non-null task_actions

In [12]:
taskactions = job120['task_action'].dropna().unique().tolist()
taskactions

['"gather_facts"',
 '"include_vars"',
 '"set_fact"',
 '"include_tasks"',
 '"package"',
 '"command"',
 '"lineinfile"',
 '"stat"',
 '"template"',
 '"service"']

### 1. For each task/module, get all the events
### 2. Group the events using the `uuid` and `parent_uuid` relation between the events
### 3. Once a group has been identified get the Time delta between the main event that spawned other events and the child event with the maximum value for Timestamp
### 4. Print all the tasks along with their time_elapsed values

### Note: Some tasks/modules would be called multiple times. For E.g. `include_vars`, `set_facts`

In [13]:
tot = 0
for task in taskactions:
    taskdf = job120.loc[job120['task_action'] == task]
    for index, row in taskdf.iterrows():
        rows = taskdf.loc[taskdf['parent_uuid'] == row['uuid']]
        if len(rows) > 0:
            maxtime = rows['time_created'].idxmax()
            te = time_elapsed(row['time_created'], rows.loc[maxtime]['time_created'])
            print(task, te)
            tot = tot + te
    
tot

"gather_facts" 2.431331
"include_vars" 0.238953
"include_vars" 0.183963
"include_vars" 0.254569
"include_vars" 0.233633
"set_fact" 0.304176
"set_fact" 0.286374
"include_tasks" 0.187121
"include_tasks" 0.486407
"package" 11.720262
"command" 1.105355
"command" 0.201508
"lineinfile" 0.91692
"stat" 0.187378
"stat" 0.201003
"template" 1.782343
"service" 1.795749
"service" 2.225288


24.742333

### Compare the above Time in seconds with the `elapsed` time in `unified_jobs_table.csv` for Job120

In [14]:
unified_jobs_df = pd.read_csv('unified_jobs_table.csv')
unified_job120 = unified_jobs_df[unified_jobs_df.id == jobid]
unified_job120.elapsed

1    52.991
Name: elapsed, dtype: float64

In [15]:
print(jobid, unified_job120.elapsed, tot)

120 1    52.991
Name: elapsed, dtype: float64 24.742333


### Note the difference between `elapsed` time in `unified_jobs_table.csv` = 52.991 Vs. The total of time calculated for each task during the job execution = 24.742